File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
from collections import deque

class Solution:
    def longestSubarray(self, nums, limit):
        maxq = deque()
        minq = deque()
        n = len(nums)
        j = 0
        ans = 0

        for i in range(n):
            while maxq and nums[i] > nums[maxq[-1]]:
                maxq.pop()
            maxq.append(i)
            while minq and nums[i] < nums[minq[-1]]:
                minq.pop()
            minq.append(i)
            if nums[maxq[0]] - nums[minq[0]] > limit:
                if nums[j] == nums[maxq[0]]:
                    maxq.popleft()
                if nums[j] == nums[minq[0]]:
                    minq.popleft()
                j += 1
            ans = max(ans, i - j + 1)
        return ans

File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java
import java.util.Deque;
import java.util.LinkedList;

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> maxq = new LinkedList<>();
        Deque<Integer> minq = new LinkedList<>();
        int n = nums.length;
        int j = 0;
        int ans = 0;

        for (int i = 0; i < n; i++) {
            while (!maxq.isEmpty() && nums[i] > nums[maxq.peekLast()]) {
                maxq.pollLast();
            }
            maxq.offerLast(i);
            while (!minq.isEmpty() && nums[i] < nums[minq.peekLast()]) {
                minq.pollLast();
            }
            minq.offerLast(i);
            if (nums[maxq.peekFirst()] - nums[minq.peekFirst()] > limit) {
                if (nums[j] == nums[maxq.peekFirst()]) {
                    maxq.pollFirst();
                }
                if (nums[j] == nums[minq.peekFirst()]) {
                    minq.pollFirst();
                }
                j++;
            }
            ans = Math.max(ans, i - j + 1);
        }
        return ans;
    }
}

